jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/AbstractAdvanceRuntimeDataServiceImpl.java

```diff
@@ -0,0 +1,306 @@
+
+    public AbstractAdvanceRuntimeDataServiceImpl() {
+        QueryManager.get().addNamedQueries("META-INF/Servicesorm.xml");
+    }
+    public void setCommandService(TransactionalCommandService commandService) {
+        this.commandService = commandService;
+    }
+
+    public void setEmf(EntityManagerFactory emf) {
+        this.emf = emf;
+    }
+
+
+    public List<org.jbpm.services.api.model.ProcessInstanceWithVarsDesc> queryProcessByVariables(Map<String, Object> attributes,
+                                                                                                 Map<String, Object> variables,
+                                                                                                 int processType,
+                                                                                                 String varPrefix,
+                                                                                                 QueryContext queryContext) {
+        EntityManager entityManager = emf.createEntityManager();
+
+        // first step is to filter the data creating a derived tables and pivoting var - rows to columns (only the variables we are interested to filter)
+        StringBuilder where = new StringBuilder();
+        StringBuilder derivedTables = new StringBuilder();
+        for (Map.Entry<String, Object> var : variables.entrySet()) {
+            String alias = "TABLE_" + var.getKey();
+            String varName = var.getKey();
+
+            derivedTables.append("LEFT JOIN (" +
+                                 "SELECT A1.processId, A1.processInstanceId, A1.variableId AS NAME, A1.value AS VALUE \n" +
+                                 "FROM VariableInstanceLog A1 \n" +
+                                 "LEFT JOIN VariableInstanceLog A2 ON A1.processId = A2.processId AND A1.processInstanceId = A2.processInstanceId AND A1.variableInstanceId = A2.variableInstanceId AND A2.log_date < A1.log_date  \n" +
+                                 "WHERE A2.log_date IS NULL \n" +
+                                 ") " + alias + " ON " + alias + ".processId = pil.processId AND " + alias + ".processInstanceId = pil.processInstanceId \n");
+
+            addWhereExpression(where);
+            where.append(alias + ".NAME = :NAME_" + varName + " AND " + alias + ".VALUE= :VALUE_" + varName + "\n");
+        }
+
+
+        for (Map.Entry<String, Object> entry : attributes.entrySet()) {
+            addWhereExpression(where);
+            where.append(" pil." + entry.getKey() + " = :ATTR_" + entry.getKey() + " ");
+        }
+
+        addWhereExpression(where);
+        where.append(" pil.processType = :processType ");
+
+        String procSQLString = "SELECT DISTINCT pil.processInstanceId FROM ProcessInstanceLog pil \n " + derivedTables + where + " ORDER BY pil.processInstanceId ASC ";
+        procSQLString += (queryContext.getCount() > 0) ? " LIMIT " + queryContext.getCount() + " OFFSET " + queryContext.getOffset() + " \n" : "";
```

Comment:
This native query is failing when using an Oracle database: **ORA-00933: SQL command not properly ended**

Seems it could be related to the pagination part of the SQL statement.